### PR TITLE
Add extensions in QueryError

### PR DESCRIPTION
**Pull request by ivanp**
Added "extensions" field in QueryError to conform with [Apollo Server errors](https://www.apollographql.com/docs/apollo-server/features/errors.html).

### CODE REVIEW

**Comment by pavelnikolov** [comment]
What problem does this solve? I'm very interested in adding extensions to this library but I don't see the value of this commit 🤔

**Comment by ivanp** [comment]
I am trying to write a custom error response according to the [GraphQL error spec](https://facebook.github.io/graphql/June2018/#sec-Errors), so i can write something like this:

```{
  "errors": [
    {
      "message": "Name for character with ID 1002 could not be fetched.",
      "locations": [ { "line": 6, "column": 7 } ],
      "path": [ "hero", "heroFriends", 1, "name" ],
      "extensions": {
        "code": "CAN_NOT_FETCH_BY_ID",
        "timestamp": "Fri Feb 9 14:33:09 UTC 2018"
      }
    }
  ]
}```

**Comment by smithaitufe** [comment]
According to the spec, the extensions property must be a map or a hash. It says that when it is serialized to json, it should be an object.

**Comment by pavelnikolov** [comment]
Can you, please, add unit tests?

**Comment by ivanp** [comment]
I was thinking, what if we make the ResolverError field outputs to json?

```
type QueryError struct {
	Message       string                 `json:"message"`
	Locations     []Location             `json:"locations,omitempty"`
	Path          []interface{}          `json:"path,omitempty"`
	Rule          string                 `json:"-"`
	ResolverError error                  `json:"extensions,omitempty"`
}
```

That way if the resolver returns a custom error with tags it will be displayed in extensions field.

File: errors/errors.go [hunk 1]
```diff
@@ -9,7 +9,7 @@ type QueryError struct {
 	Locations     []Location    `json:"locations,omitempty"`
 	Path          []interface{} `json:"path,omitempty"`
 	Rule          string        `json:"-"`
-	ResolverError error         `json:"-"`
+	ResolverError error         `json:"extensions,omitempty"`
```
Comment (pavelnikolov) [inline errors/errors.go]: Shouldn't we name this field `Extensions` with type `map[string]interface{}`. This way it can still contain an error as well as other fields. For example, it might be a validation error and you might include the error(s) as well as the invalid input field(s) or you might want to include request ID, trace ID or other fields that are not directly related to the error 🤔

Comment (smithaitufe) [reply errors/errors.go]: I agree with @pavelnikolov  that the name should be `Extensions` and the type should be a `map[string]interface{}`

**Comment by ivanp** [comment]
OK I have reverted to the previous commit to add `Extensions` as `map[string]interface{}`.

File: errors/errors.go [hunk 2]
```diff
@@ -5,11 +5,12 @@ import (
...
-	Locations     []Location    `json:"locations,omitempty"`
-	Path          []interface{} `json:"path,omitempty"`
-	Rule          string        `json:"-"`
-	ResolverError error         `json:"-"`
+	Message       string                 `json:"message"`
+	Locations     []Location             `json:"locations,omitempty"`
+	Path          []interface{}          `json:"path,omitempty"`
+	Rule          string                 `json:"-"`
+	ResolverError error                  `json:"-"`
+	Extensions    map[string]interface{} `json:"extensions,omitempty"`
```
Comment (pavelnikolov) [inline errors/errors.go]: I'm wondering if the `Extensions` field should be used in the `Error()` method as well... 🤔

File: graphql_test.go [hunk 3]
```diff
@@ -304,6 +323,45 @@ func TestNilInterface(t *testing.T) {
 	})
 }
 
+func TestErrorWithExtension(t *testing.T) {
+	err := withExtensionError{
+		Code:    "NotFound",
+		Message: "This is not the droid you are looking for",
```
Comment (pavelnikolov) [inline graphql_test.go]: The test is named `TestErrorWithExtension`, however the error doesn't have an extension field set. Was this indetntional? Can you, please, add some more test cases? Then I'll be happy to merge this PR.

Comment (pavelnikolov) [reply graphql_test.go]: Maybe you need to add cases both with and without the extension field 🤔

Comment (pavelnikolov) [reply graphql_test.go]: Well you can just return:
```go
err := errors.QueryError{
    Message: "error message goes here",
    Extensions: map[string]interface{}{"key1": "val1", "key2": "val2"},
}

return nil, err
```

Comment (pavelnikolov) [reply graphql_test.go]: I guess we have to return some kind of a custom error that implements this interface (this interface will not be part of the library):
```go
type ResolverError interface {
    error
    Extensions() map[string]interface{}
}

```

Then at the top of `internal/exec/exec.go` define this interface:
```go
type extensionser interface {
	Extensions() map[string]interface{}
}
```

and later in `internal/exec/exec.go:190` the code will look like:
```go
if f.field.HasError && !callOut[1].IsNil() {
    resolverErr := callOut[1].Interface().(error)
    err := errors.Errorf("%s", resolverErr)
    err.Path = path.toSlice()
    err.ResolverError = resolverErr
    if ex, ok := callOut[1].Interface().(extensionser) {
        err.Extensions = ex.Extensions()
    }
    return err
}
```

This way it will be up to the user to define any custom error type (e.g. `AuthenticationError`, `ForbiddenError`, `ValidationError`, `ServerError`, etc.) 
Ideally, this will also need to be documented with another example.

**Comment by pavelnikolov** [comment]
Thank you for your contribution.
You might as well consider sending an example with custom errors 🤔

**Comment by ivanp** [comment]
Thanks! Got it, I'll send you some examples.

**Comment by pavelnikolov** [comment]
The example needs to be added as a testable example in the `errors` package so that it appears here: https://godoc.org/github.com/graph-gophers/graphql-go/errors
